### ISSUE TITLE
implement dialog with download options for folder downloads

### ISSUE THREAD

**Issue by mbarchein** [nicotine-plus/nicotine-plus#1659]
Hello. Thanks for the excellent Nicotine+ software. I have a feature request. 

This is one of my favourite SoulseekQT features not present in Nicotine+. When I download a folder from another user, the UI allows me to change the target folder name before starting the download. This is different from selecting the target container folder. It could be useful when the folder name is, for example, "CD1". I would like it to store that folder on my computer as "Artist name - year - album CD1".

Also, SoulseekQT allows selecting which files of the folder will be downloaded in case you don't want to download every file present in the folder. That's also a neat feature not present in Nicotine+

Thank you very much.

**Comment by slook** [comment]
The dialog looks like this https://github.com/nicotine-plus/nicotine-plus/issues/1717#issuecomment-974716135 and would allow for a range of enhancements to be added for when control over downloads is required, this is a very nice idea.

**Comment by mathiascode** [comment]
How does only showing the dialog when pressing the `Download Folder(s) To...` menu item sound? The idea would be that the `Download Folder(s)` item remains an option if you quickly want to grab a few complete folders, while `Download Folder(s) To...` provides all the fine-grained options that might be useful, such as selecting the container folder, folder name, which files to download etc.

**Comment by ngs428** [comment]
Thanks for working to add this.  It will be a great feature!

**Comment by catkryton** [comment]
Hi, I have implemented this functionality and it works well. How can I contribute this to the project?
![image](https://github.com/nicotine-plus/nicotine-plus/assets/135138188/a3ae8d36-759b-4685-b8fe-1aa2332a814c)

**Comment by slook** [comment]
> right click a file and select 'download folder'

I think we still want to retain the option of making a complete folder download without a dialog prompt, see https://github.com/nicotine-plus/nicotine-plus/issues/1659#issuecomment-1312742475.

> set the path if you would like it to be different

Implement a FolderChooser() widget for this, see the preferences dialog module DownloadPage().download_folder_button for an example of this.

> the files are added to the transfer list

An ToggleButton option to add as Paused with locally queued status set is also desired see https://github.com/nicotine-plus/nicotine-plus/issues/1717 to delay the transfer for later.

> How can I contribute this to the project?

Fork the repo, create a new branch and then submit a PR (pull request).

**Comment by slook** [comment]
1. Agree that measures are needed to avoid accidental folder downloads, with the balance of convenience. Hence that is the reason why the double-click action was disabled for folder rows. You might want to consider enabling it again if the new dialog shall always ask for confirmation anyway.

2. Good point. The dialog is intended to replace the standard folder chooser dialog, so it would be silly to spawn a dialog ontop, when it is probable that the intended destination folder won't exist yet anyway. A manual text entry is desirable.

3. No, you're right a dialog shall not be required for the Browse Shares interface.

> The Queue Paused should be easy to add.

I don't know how we intend to approach the invocation of the transfers module in this case. I have a feeling it will lead to the discovery of old bugs that will lead to a regression, because the transfers module gets upset easily.

**Comment by mathiascode** [comment]
Thanks for working on this, looks good so far. One point:
- Could you add the Download button to the dialog (see widgets/dialogs.py, `buttons_end`)? My plan is to eventually add support for downloading many folders at once, which will use a similar back/forward navigation style as the Setup Assistant. The containing folder selector will remain on all pages, while the rest of the content changes.

**Comment by slook** [comment]
> The containing folder selector will remain on all pages, while the rest of the content changes.

Surely the destination folder initial suggested value will be different for each selected source folder?

**Comment by slook** [comment]
That would be one way of doing it. Alternatively, it might be better to include a unique path entry widget for each page.

In @catkryton 's screenshot it appears that the destination folder name is acquired from the remote source.

**Comment by mathiascode** [comment]
Sorry, I won't be able to get this done for 3.3.0. I want to work on it eventually, but right now I need to spend the time I have left finishing up the 3.3.0 release and fixing any remaining bugs.

**Comment by mathiascode** [comment]
I've started working on this feature and thinking about the design again. I think we can do better than SoulseekQt's dialog and its limitations (no file sizes, one folder at a time). This is what I've come up with so far:

A single menu item for downloading in search results (having four of them is too much...):
![menu](https://github.com/nicotine-plus/nicotine-plus/assets/8754153/aaf347dd-f2a7-4e39-9a2f-c2078c886c07)

Download dialog with the following features:
- Set a different download location than the default
- Choose if files should be saved to subfolders or not (disabling this option behaves like the old `Download files` context menu item)
- A treeview of all selected folders and any additional files that were fetched for them. The names of subfolders can be changed here.
- If downloading a single folder, all files will be selected by default. If downloading many folders/files, only files from search results will be selected by default. Will have to do some testing to see if this is a good idea in practice or not, can probably refine the behavior here further.
- More features can be added to the dialog in the future (audio quality/duration? file properties dialog?)

![download dialog](https://github.com/nicotine-plus/nicotine-plus/assets/8754153/78ccb8b7-7bb3-4c2a-80b8-3c0debf62695)

![rename dialog](https://github.com/nicotine-plus/nicotine-plus/assets/8754153/c887e56f-8b48-4052-a05c-baa49076b7d3)

**Comment by slook** [comment]
Very nice.

Inevitably some users will want to "Always download files to default download location" so such an option should be provided to either suppress the dialog or the old menu item might be retained for quick access.

The rename folder dialog doesn't need to display the entire full path of the old name, since that information is irrelevant in this context.

The folder checkboxes in the tree view should show a 'partially checked' tick (normally filled with grey) if some but not all containing files are selected.

Clicking on a subfolder checkbox could ideally have 3 states:
- none selected;
- initially selected in that subfolder;
- all selected.

**Comment by mathiascode** [comment]
> Inevitably some users will want to "Always download files to default download location" so such an option should be provided to either suppress the dialog or the old menu item might be retained for quick access.

I think waiting for feedback from users before reintroducing ways to skip the dialog would be the best idea. Then we can find out which functions people actually use.

> The rename folder dialog doesn't need to display the entire full path of the old name, since that information is irrelevant in this context.

I initially included it to provide more context for folder names such as "Disc 1" or "CD 1". I've since changed the behavior to also include the parent folder if the subfolder name is too short, so we can probably remove the full path.

> The folder checkboxes in the tree view should show a 'partially checked' tick (normally filled with grey) if some but not all containing files are selected.
> 
> Clicking on a subfolder checkbox could ideally have 3 states:
> 
> * none selected;
> * initially selected in that subfolder;
> * all selected.

I'll have to see what's possible with the tree view checkboxes.

> Often the list of files is long, so it cannot always be visibly seen how many items are selected. It is essential to see the total size of the selected items (number/size/duration), as discussed in #2697, perhaps in a status bar label, or, in the title bar "Download n Files (x MiB)".

Title bar would probably be the best location. The current static title is redundant anyway.

**Comment by slook** [comment]
> I've since changed the behavior to also include the parent folder name if the subfolder name is too short

This might not be desirable because arbitrarily hacking the folder names might get in the way of the ability to locally recreate the remote folder structure.

**Comment by mathiascode** [comment]
I've restored the "Download files" option to avoid showing a dialog every time someone double-clicks a file.

> This might not be desirable because arbitrarily hacking the folder names might get in the way of the ability to locally recreate the remote folder structure.

What do you suggest? Currently we end up with a bunch of "CD x" folders, with no indication of what the original album name is. In the worst case there are multiple folders with this identical name when downloading from search results.

**Comment by slook** [comment]
> What do you suggest?

Create a containing parent folder with the name from the level below. In other words, retain the path separator.

**Comment by Rings4** [comment]
> > What do you suggest?
> 
> Create a containing parent folder with the name from the level below. In other words, retain the path separator.

+1 for this. I just started using nicotine+ and when I double click a file when browsing someone's shares, the file downloads to my completed folder without any subfolder. In Soulseek, it would download to its own subfolder, which is usually the album name etc.

**Comment by slook** [comment]
If you are able to run https://github.com/nicotine-plus/nicotine-plus/pull/3024 development so far can be tested out. To do this you could either clone the `search-download-dialog` development branch in Git or install one of the artifacts from the relevant actions packaging:

https://github.com/mathiascode/nicotine-plus/tree/search-download-dialog
https://github.com/nicotine-plus/nicotine-plus/actions/runs/9428931585?pr=3024

Thank you for your feedback.

**Comment by Rings4** [comment]
I haven't had time to try the above yet, but I noticed some inconsistent behaviour on latest stable release that I wanted to outline as it may greatly facilitate fixing this issue.

My friend downloaded a file which put the file in the root of his download folder, without a subfolder. However, I then sent a file to my friend and that file downloaded in its own subfolder.

So the upload file feature is correctly placing downloaded files in their proper subfolders on the recipient's end.

**Comment by mathiascode** [comment]
> I haven't had time to try the above yet, but I noticed some inconsistent behaviour on latest stable release that I wanted to outline as it may greatly facilitate fixing this issue.
> 
> My friend downloaded a file which put the file in the root of his download folder, without a subfolder. However, I then sent a file to my friend and that file downloaded in its own subfolder.

It's intentional behavior. In many cases, people just want to download individual files to the root of their download folder, and Nicotine+ has behaved as such since the beginning. SoulseekQt has the option to also store individual files in subfolders, similar to downloading a whole folder.

Since we don't know if received files are complete folders or not, they are always placed in a subfolder.

What are your reasons for wanting to store individual files in subfolders? Do you usually download single files, or a few files from a folder?

**Comment by Rings4** [comment]
> > I haven't had time to try the above yet, but I noticed some inconsistent behaviour on latest stable release that I wanted to outline as it may greatly facilitate fixing this issue.
> > My friend downloaded a file which put the file in the root of his download folder, without a subfolder. However, I then sent a file to my friend and that file downloaded in its own subfolder.
> 
> It's intentional behavior. In many cases, people just want to download individual files to the root of their download folder, and Nicotine+ has behaved as such since the beginning. SoulseekQt has the option to also store individual files in subfolders, similar to downloading a whole folder.
> 
> Since we don't know if received files are complete folders or not, they are always placed in a subfolder.
> 
> What are your reasons for wanting to store individual files in subfolders? Do you usually download single files, or a few files from a folder?

I have never found myself not wanting to save a file in a subfolder. Sometimes I'm downloading several versions of the same file when I'm searching for a specific version that has the same name, would be very annoying having to rename or organize each one before being able to download another.

And sometimes I simply want a song and not an entire folder. I still want that to be downloaded to its own folder. I often download stuff and then forget about them since downloads take forever sometimes, so I'd have no idea what they are when I go back later on and notice a bunch of random unorganized files.

Basically, I'll never want any downloads to simply to go the root folder so I'd love for an option that would sort all downloads in their respective subfolder named after the host's subfolder.

**Comment by Rings4** [comment]
Another inconsistent behaviour I noticed is the fact that sending a folder containing subfolders to a user will only send the subfolders. 

For example, I have a folder "Label [FLAC]" with subfolders "Album 1 [FLAC]" and "Album 2 [FLAC]". 

Right clicking "Label [FLAC]" and clicking "Upload folder and subfolders to..." and selecting my friend only sends him "Album 1 [FLAC]" and "Album 2 [FLAC]".

However, if my friend right clicks "Label [FLAC]" and clicks "Download folder and subfolders", it gets "Label [FLAC]" with "Album 1 [FLAC]" and "Album 2 [FLAC]" inside.

**Comment by mathiascode** [comment]
> It's obviously working properly for my friend when he downloads from me, but not when I upload to him. I would like the same behaviour to be used for sending files.

This requires a change in the client on the receiving end, but it shouldn't be to hard to add in Nicotine+ at least. Basically check the common folders of the sent file paths, and create them instead of only the immediate parent folder.

**Comment by AverageHoarder** [comment]
Another idea to avoid/reduce the annoying case of downloading multiple "CD1"/"scans"/"artwork" folders from the same user whose files end up in the same folder (possibly overwriting one another) would be to perform a check to see if the currently selected subfolder already exists and display a warning in the UI if it does (or even make the user confirm the warning).

Concerning the pre-selection behavior I think it would make sense to change it up depending on what's selected in nicotine when the context menu is called.
If only a couple songs within one folder are selected, preselect only these.
If a couple of songs from multiple folders are selected, preselect only these.
If a folder is selected: preselect everything in it (minus those files that match download filters like .exe etc.)
If multiple folders are selected: preselect everything in these (minus those files that match download filters like .exe etc.)

Other than that, I think the design preview you've shared, @mathiascode looks good.
One thing that might help would be to add an indication that the download folder has been renamed (to show which have been set manually when downloading multiple). Perhaps by making the respective folder name cursive: _Christmas! (2014)_

**Comment by slook** [comment]
The new download dialog is now available in Nicotine+ 3.4.0.dev1 please can everybody try it and let us know how you get on...
https://nicotine-plus.org/doc/TESTING.html

**Comment by srodillof** [comment]
is there way to have this feature on older versions? I can't upgrade past Nicotine+ 3.3.6 because I have Mac OS 12 (updates are for mac os 13 onwards)

**Comment by slook** [comment]
Unfortunately we cannot backport any features beyond the head of the previous branch, which is at 3.3.10, and anyway this feature is still experimental and is way too large, unfortunately.

Also see https://github.com/nicotine-plus/nicotine-plus/discussions/3338 about MacPorts.

**Comment by StodgyWaif** [comment]
> The new download dialog is now available in Nicotine+ 3.4.0.dev1 please can everybody try it and let us know how you get on... https://nicotine-plus.org/doc/TESTING.html

Hello.
I've installed the latest testing version and the folder dialog is great.
However the download filter is not being applied so I still needed to uncheck the unwanted files.

**Comment by slook** [comment]
It will be applied afterwards, it will show up in the Downloads list with "Filtered".

**Comment by StodgyWaif** [comment]
I see. Yes it's working great.
Thank you for clarifying.

**Comment by ngs428** [comment]
> The new download dialog is now available in Nicotine+ 3.4.0.dev1 please can everybody try it and let us know how you get on... https://nicotine-plus.org/doc/TESTING.html

I am late to the testing, but it is working as intended, thanks!  Thanks @catkryton for the assist!  Nice work!!!
